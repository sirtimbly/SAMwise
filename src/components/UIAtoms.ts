import * as Mithril from 'mithril';

let m = Mithril;

/* 
Atoms are very generic and abstract so that they can be reused in the most places.
They know nothing about where they will be implemented, the shape of the model, or 
the details of the action functions they call.
*/


/**
 * UIProgressBar
 * @param value 
 * @param max 
 * @param label 
 */
export const UIProgressBar = function(value:number, max:number, label?:string) {
    return m('div.guage-item.flex.px2', [
        m('progress.col.col-4.mr2', {
            value: (value),
            max: max
        }),
        label ? m('span.bold', label) : ''
    ])
}

/**
 * UITab
 * @param label string which displays in the tab
 * @param active if the tab is currently selected
 * @param action the click action function to call, will be passed nothing
 */
export const UITab = function(label:string, active:boolean, action:()=>void) {
    return m("button.btn.col-2.mt1.mr1.btn-large.rounded-top", {
        class: active ? 'bg-blue' : 'bg-light-gray',
        onclick: (e) => {action()}
    }, label)
}