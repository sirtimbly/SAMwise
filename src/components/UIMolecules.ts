import * as m from 'mithril';

// let m = Mithril;

/*
Molecules are more specific than Atoms and have some more complicated layout or logic
and we expect them to be somewhat less reusable than Atoms but still not referring 
to any specific actions or models from the app implementation.
*/

/**
 * Slider is a component with heading and a range input that changes either fulltime or parttime values
 */
export const UISlider = function (label:string, count:number, changeAction:(value:number)=>void, max:number = 100) {
    let uName = "range_"+encodeURIComponent(label);
    return m("div.flex.bg-light-gray.p2.m1.flex-reverse", 
    [    
        m("input.col-8", {
            type: "range",
            min: "0",
            max: max || 100,
            name: uName,
            value: count,
            onchange: (e) => {changeAction(e.target.value)}
        }),
        m("label.title.col-4.px2", 
        {for: uName},
        count + " " + label + " shifts assigned.")
    ])
}
