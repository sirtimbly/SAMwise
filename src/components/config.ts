import * as mithril from "mithril";

