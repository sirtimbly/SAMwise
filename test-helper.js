"use strict";
exports.__esModule = true;
var htmltidy = require("tidy-html5");
var m = require("mithril");
exports.defaultHtmlTidyOptions = {
    "show-body-only": true,
    "drop-empty-elements": false,
    "doctype": "omit",
    "indent": true,
    "quiet": true,
    "show-warnings": false,
    // Recognize SVG tags:
    "new-blocklevel-tags": ["svg", "defs"],
    "new-inline-tags": ["path", "polyline", "line", "polygon"]
};
exports.tidy = function (vnodes, htmltidyOptions) {
    if (htmltidyOptions === void 0) { htmltidyOptions = exports.defaultHtmlTidyOptions; }
    var htmlElement = document.createElement("div");
    m.render(htmlElement, vnodes);
    var html = htmlElement.innerHTML;
    return htmltidy.tidy_html5(html, htmltidyOptions);
};
