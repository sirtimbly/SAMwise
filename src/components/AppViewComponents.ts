import * as m from 'mithril';
// import Mithril from 'mithril';
import { App, State, Model, Action } from '../samwise';
import { UITab, UIProgressBar } from './UIAtoms';
import { UISlider } from './UIMolecules';
import { AssignmentModel, AssignmentProps } from '../models/assignment_model';
/* ============================================ */
/**
 * App View Components (these functions make the actual UI layer updates). 
 * They are implementation specific and tightly coupled to the actions above from the app.actions object.
 * + edit
 *  - slider
 * + review
 */
interface VState {} 


export class AppViewComponents {
    app:App;
    constructor (app:App) {
        this.app = app;
    }
    /**
     * Tabs component is coupled to the state model so we define it here instead of in the UIMolecules file
     */
    public Tabs = {
        view: (vnode) => {
            return m("div.flex.flex-row.mx3.mt1", [
                UITab('Widget Production Plan', vnode.attrs.isEditing, this.app.actions.switchToEdit),
                UITab('Review', !vnode.attrs.isEditing, this.app.actions.switchToReview)
            ])
        }
    } as m.Component<AssignmentProps,VState> & VState 

    

    /**
     * Edit Component includes multiple sliders
     */
    public Edit = {
        view: (vnode) => {
 
            console.log(vnode.attrs)
            let ftAvail = vnode.attrs.requested - vnode.attrs.parttime;
            let ptAvail = vnode.attrs.requested - vnode.attrs.fulltime;
            return m('div.border.border-blue.bg-white.rounded.p2.mx2.flex.flex-column', 
            [
                UIProgressBar(ftAvail, vnode.attrs.requested, "Big Widget slots: " + ftAvail),
                UIProgressBar(vnode.attrs.fulltime, vnode.attrs.requested),
                UISlider('Big Widget', vnode.attrs.fulltime, (val) => this.app.actions.changeAction('fulltime', val), vnode.attrs.requested),
                
                UIProgressBar(ptAvail, vnode.attrs.requested, "Small Widget slots: " + ptAvail ),
                UIProgressBar(vnode.attrs.parttime, vnode.attrs.requested),
                UISlider('Small Widget', vnode.attrs.parttime, (val) => this.app.actions.changeAction('parttime', val), vnode.attrs.requested),

                m(this.Footer, vnode.attrs)
            ]);
        }
    } as m.Component<AssignmentProps, VState>
    
    
    /**
     * Footer shows schedule versus total and a checkmark icon
     */
    public Footer = {
        view: (vnode) => {
            
            let matchClass = vnode.attrs.isValid ? 'icon-ok' : '';
            return m('div.p2.m1',
                {
                    class: vnode.attrs.isValid ? 'bg-green' : 'bg-red'
                },
                'Widget Quota: ' + vnode.attrs.total + ' of ' + vnode.attrs.requested, m('span', { class: matchClass }))
        }
    } as m.Component<AssignmentProps, VState>

    /**
     * Review Component 
     */
    public Review = {
        view: (vnode) => {
           
            return m("div.border.mx2", [
                m("table.table", [
                    m("tbody", [
                        m("tr", [
                            m("td", "Big Widget Slots"),
                            m("td", vnode.attrs.fulltime)
                        ]),
                        m("tr", [
                            m("td", "Small Widget Slots"),
                            m("td", vnode.attrs.parttime)
                        ]),
                        m("tr.text-bold.bg-blue", [
                            m("td", "Total"),
                            m("td", [
                                m(this.Footer, vnode.attrs)
                            ])
                        ])
                    ])
                ]),
                m('button.btn.btn-lg.m3',
                {
                    class: vnode.attrs.isValid ? 'bg-blue': 'bg-gray',
                    disabled: !vnode.attrs.isValid ? 'disabled' : '',
                    onclick: (e) => this.app.actions.saveData()
                },
                'Save')
            ])
        }
        
    } as m.Component<AssignmentProps, VState>

    
}