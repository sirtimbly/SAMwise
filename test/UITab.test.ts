import * as m from "mithril";
import {tidy} from '../test-helper'

import {UITab} from '../src/components/UIAtoms'
import * as mq from 'mithril-query'

describe("UITab", () => {
  it("should have text", () => {
    const cmp = UITab('hello tab', true, () => {});
    const output = mq(cmp);
    expect(output.contains('hello tab')).toBeTruthy();

    const html = tidy(cmp);
    expect(html).toMatchSnapshot();
  });
  it ("should be clickable", () => {
    let actionTriggered = false;
    const cmp = UITab('hello tab', actionTriggered, () => {actionTriggered = true});
    const output = mq(cmp);
    output.click('button')
    expect(actionTriggered).toBeTruthy();
  })
});