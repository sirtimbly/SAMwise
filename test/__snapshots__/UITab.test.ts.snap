// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UITab should have text 1`] = `
"<button class=\\"btn btn-lg mt1 mr1 rounded-top bg-blue\\">hello
tab</button>
"
`;
